MoreLinq.Test/TestingSequence.cs

```diff
@@ -51,59 +63,268 @@ public enum Options
+            if (_enumerationCount > 0)
+                Assert.That(_disposedCount, Is.EqualTo(_enumerationCount), ExpectedDisposal);
         }
 
         public IEnumerator<T> GetEnumerator()
         {
-            if (!IsReiterationAllowed)
-                Assert.That(_sequence, Is.Not.Null, "LINQ operators should not enumerate a sequence more than once.");
-
-            Debug.Assert(_sequence is not null);
+            Assert.That(_enumerationCount, Is.LessThan(_maxEnumerations), TooManyEnumerations);
+            Assert.That(_enumerationCount, Is.EqualTo(_disposedCount), SimultaneousEnumerations);
+            _enumerationCount++;
 
             var enumerator = _sequence.GetEnumerator().AsWatchable();
-            _disposed = false;
+            var disposed = false;
             enumerator.Disposed += delegate
             {
-                Assert.That(_disposed, Is.False, "LINQ operators should not dispose a sequence more than once.");
-                _disposed = true;
+                if (!disposed)
+                {
+                    _disposedCount++;
+                    disposed = true;
+                }
+                else if (!_options.HasFlag(Options.AllowRepeatedDisposals))
+                {
+                    Assert.Fail(TooManyDisposals);
+                }
             };
+
             var ended = false;
             enumerator.MoveNextCalled += (_, moved) =>
             {
-                Assert.That(ended, Is.False, "LINQ operators should not continue iterating a sequence that has terminated.");
+                Assert.That(disposed, Is.False, MoveNextPostDisposal);
+                if (!_options.HasFlag(Options.AllowRepeatedMoveNexts))
+                    Assert.That(ended, Is.False, MoveNextPostEnumeration);
+
                 ended = !moved;
                 MoveNextCallCount++;
             };
 
-            if (!IsReiterationAllowed)
-                _sequence = null;
+            enumerator.GetCurrentCalled += delegate
+            {
+                Assert.That(disposed, Is.False, CurrentPostDisposal);
+                Assert.That(ended, Is.False, CurrentPostEnumeration);
+            };
 
             return enumerator;
         }
 
         IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+    }
+
+    [TestFixture]
+    public class TestingSequenceTest
+    {
+        [Test]
+        public void TestingSequenceShouldValidateDisposal()
+        {
+            static IEnumerable<int> InvalidUsage(IEnumerable<int> enumerable)
+            {
+                var _ = enumerable.GetEnumerator();
+
+                yield break;
+            }
+
+            static void Act()
+            {
+                using var xs = Enumerable.Range(1, 10).AsTestingSequence();
+                InvalidUsage(xs).Consume();
+            }
+
+            Assert.That(
+                Act,
+                Throws.InstanceOf<AssertionException>()
+                    .With.Message.Contains(ExpectedDisposal));
```

Comment:
All such asserts read simpler on one line instead being spread on 4:

```suggestion
            Assert.That(Act, Throws.InstanceOf<AssertionException>().With.Message.Contains(ExpectedDisposal));
```

Better yet, since they all have the same structure, having a helper like:

```c#
static void AssertError(TestDelegate act, string message) =>
    Assert.That(act, Throws.InstanceOf<AssertionException>()
                           .With.Message.Match(@"^\s*" + Regex.Escape(message)));
```

will make it even simpler:

```c#
AssertError(Act, ExpectedDisposal);
```

Reply:
Agree to disagree on the multiple lines.

Common method addressed in b34d1f2